stl/inc/algorithm

```diff
@@ -4970,6 +4970,94 @@ _FwdIt reverse_copy(_ExPo&&, _BidIt _First, _BidIt _Last, _FwdIt _Dest) noexcept
     _REQUIRE_PARALLEL_ITERATOR(_FwdIt);
     return _STD reverse_copy(_First, _Last, _Dest);
 }
+
+#ifdef __cpp_lib_concepts
+namespace ranges {
+    // ALIAS TEMPLATE reverse_copy_result
+    template <class _In, class _Out>
+    using reverse_copy_result = in_out_result<_In, _Out>;
+
+    // VARIABLE ranges::reverse_copy
+    class _Reverse_copy_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <bidirectional_iterator _It, sentinel_for<_It> _Se, weakly_incrementable _Out>
+            requires indirectly_copyable<_It, _Out>
+        constexpr reverse_copy_result<_It, _Out> operator()(_It _First, _Se _Last, _Out _Result) const {
+            // clang-format on
+            _Adl_verify_range(_First, _Last);
+            auto _UFirst = _Get_unwrapped(_STD move(_First));
+            auto _ULast  = _Get_final_iterator_unwrapped<_It>(_UFirst, _STD move(_Last));
+            _Seek_wrapped(_First, _ULast);
+            _Result = _Reverse_copy_common(_STD move(_UFirst), _STD move(_ULast), _STD move(_Result));
+            return {_STD move(_First), _STD move(_Result)};
+        }
+
+        // clang-format off
+        template <bidirectional_range _Rng, weakly_incrementable _Out>
+            requires indirectly_copyable<iterator_t<_Rng>, _Out>
+        constexpr reverse_copy_result<borrowed_iterator_t<_Rng>, _Out> operator()(_Rng&& _Range, _Out _Result) const {
+            // clang-format on
+            if constexpr (common_range<_Rng>) {
+                _Result = _Reverse_copy_common(_Ubegin(_Range), _Uend(_Range), _STD move(_Result));
+                return {_RANGES end(_Range), _STD move(_Result)};
+            } else {
+                auto _ULast = _Get_final_iterator_unwrapped(_Range);
+                _Result     = _Reverse_copy_common(_Ubegin(_Range), _ULast, _STD move(_Result));
+                return {_Rewrap_iterator(_Range, _STD move(_ULast)), _STD move(_Result)};
+            }
+        }
+
+    private:
+        template <class _It, class _Out>
+        _NODISCARD static constexpr _Out _Reverse_copy_common(const _It _First, _It _Last, _Out _Result) {

```

Comment:
I just want to confirm that the reason for taking `_First` and `_Last` as type `_It` instead of `_It` and `_Se` (which I believe is what generally happens in other ranges algorithms) is because of the potential vectorization here?

Reply:
Algorithms that require `bidirectional` iterators almost universally want to process some elements from the end of the range before processing all of the elements in the middle. When you have a sentinel that isn't an iterator, there's no way to do that other than to determine the iterator value that corresponds to that sentinel and move on from there. Many (all?) of our implementations of such algorithms use a backend `_Meow_common` function that takes an iterator and sentinel of the same type (`common` as in `common_range` - a range with the same iterator and sentinel types). In a perfect world this reduces code bloat if someone calls the same algorithm with the same iterator type and different sentinel types since those calls would all end up in a common back end.

Reply:
That definitely makes sense.  I'm still getting the feel for iterator/sentinel combinations and how they work together and the different abilities they have.  Thanks!